### PR TITLE
improve legend docstring

### LINKED ISSUE #8668: handles keyword argument not documented in the help of legend
### Bug report

**Bug summary**

The following on-line documentation has an example where the `handles` keyword is used in a call to `plt.legend`: <http://matplotlib.org/users/legend_guide.html#controlling-the-legend-entries>

This keyword argument is not present in the internal function documentation.

**Matplotlib version**
  * Operating System: Linux (Xubuntu 16.04)
  * Matplotlib Version: 2.0.0
  * Python Version: 3.6

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This improves the legend docstrings:
- describe handles and labels
- add call signatures to `Axes.legend()`
- Structure and reorder `Axes.legend()` so that the most common use case is at the top.

also fixes #8668

### CODE REVIEW

**Comment by timhoffm** [comment]
Just saw the failing test, which compares if the parameter sections are identical for `axes.legend()` and `Legend()`.

I assume the intent is to keep the long parameter lists at sync. However since the signatures are actually different, this does not work easily. Either one has to break with the documenting rules and exclude parent, handles and labels from the parameter sections (probably why the docstring was in the peculiar state before I changed it), or one has to refine the test.

How should I proceed?

**Comment by jklymak** [comment]
Without looking too closely, can you keep "Parameters" as the arguments that change, and "Other Paramters" as the parameters that are shared and move the test to look at the "Other Parameters" section?  

The problem before was that the very very very long list of kwargs was not consistent between the legend docstrings.  I fixed that, and then introduced the test (#9324).  If it had been up to me, we'd have used a doc interp, but the consensus seemed to be that we don't like doc interps.  I'm not particularly happy w/ the existing solution, so other creative ideas more than welcome.  I'm not going to be terribly upset if we get rid of the test for consistency, but there should be a way to keep the kwarg documentation consistent between the three `legend`s.  See also #9414.

**Comment by timhoffm** [comment]
The workaround with "Other Parameters" should be ok in this case. Will look into it.

As of syncing docstrings, you just have the two options:
- Sync manually and monitor deviations (--> current tests).
- Any automatic mechanism at build or import-time.

I currently do not oversee the subtleties of the latter, and therefore, cannot comment here.

**Comment by timhoffm** [comment]
Docstring / test adapted.

**Comment by jklymak** [comment]
Test passed; you've got `.legend` in there somewhere, and its ambiguous between `Figure.legend` and `axes.legend`.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -283,38 +283,83 @@ def legend(self, *args, **kwargs):
...
+        or::
+
             line.set_label('Label via method')
+            line, = ax.plot([1, 2, 3])
             ax.legend()
 
         Specific lines can be excluded from the automatic legend element
         selection by defining a label starting with an underscore.
-        This is default for all artists, so calling :meth:`legend` without
+        This is default for all artists, so calling :meth:`.legend` without
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: This needs to me `.Axes.legend` I think

**Pull request merged**